### ISSUE TITLE
Custom Resources + Serverless Containers Don't work

### ISSUE THREAD

**Issue by amirhajimirsadeghi** [aws-amplify/amplify-cli#7049]
### Before opening, please confirm:

I have installed the latest version of the Amplify CLI (see above), and confirmed that the issue still persists.
I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-cli/issues?q=is%3Aissue+).
I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#bug-reports).
I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### How did you install the Amplify CLI?

npm i -g @aws-amplify/cli

### If applicable, what version of Node.js are you using?

v14.7.0

### Amplify CLI Version

4.45.2

### What operating system are you using?

Mac

### Amplify Categories

api, Not applicable

### Amplify Commands

push

### Describe the bug

Referencing Custom Resources in Serverless Containers get overwritten during `amplify push`

I followed [this guide](https://docs.amplify.aws/cli/usage/customcf#n2-modify-folder-structure-for-custom-resource) to create a custom resource and reference it in the cloudformation for my serverless container but every time I add new things and run `amplify push` the file gets overwritten with my changes no longer there. This makes it impossible to reference custom references in serverless containers

### Expected behavior

Adding custom resource references to Serverless Container using [this guide](https://docs.amplify.aws/cli/usage/customcf#n2-modify-folder-structure-for-custom-resource) should work and not overwrite the cloudformation file.

### Reproduction steps

1. enable and create a serverless container using this guide: https://docs.amplify.aws/cli/usage/containers#getting-started
2. create a custom resource using this guide: https://docs.amplify.aws/cli/usage/customcf#n2-modify-folder-structure-for-custom-resource
3. referencing the custom resource, give serverless container permission to access custom resource by following this guide: https://docs.amplify.aws/cli/usage/customcf#n2-modify-folder-structure-for-custom-resource
4. run `amplify push`
5. see that references to custom resource have been over written in serverless container cloudformation

### GraphQL schema(s)

N/A

### Log output

<details>

```
# Put your logs below this line


```

</details>

<details>

```graphql
# Put schemas below this line


```

</details>

**Comment by fr3shn3ss** [comment]
I see the same issue. No idea how the team can release this feature with such a major bug

**Comment by mattfysh** [comment]
@amirhajimirsadeghi could this be related? https://github.com/aws-amplify/amplify-cli/issues/7137

**Comment by amirhajimirsadeghi** [comment]
No, @mattfysh this is related to https://github.com/aws-amplify/amplify-cli/issues/6918 I think I double created the ticket

**Comment by josefaidt** [comment]
Hey @amirhajimirsadeghi :wave: apologies for the delay! To clarify are you looking to use outputs from your custom resource in the category resource? Currently we support taking in outputs from categories into a custom category.

**Comment by amirhajimirsadeghi** [comment]
@josefaidt I'm looking to use outputs from a custom resource in a serverless container API resource. However, in attempting to do so, the cloud formation gets overwritten and thus nothing can be used. 

Importing custom resource outputs to other resources works just fine

**Comment by josefaidt** [comment]
Hey @amirhajimirsadeghi thank you for the information! I will mark this as a feature request for the team to review

**Comment by renebrandel** [comment]
Hi @amirhajimirsadeghi - we've just published an RFC on providing containers access permissions to external resources. Would love to get your feedback on it: https://github.com/aws-amplify/amplify-cli/pull/7346

**Comment by josefaidt** [comment]
This is still a valid feature request. After creating a container-based API in our project and attempt to edit the CloudFormation template the edits are overwritten on `amplify push`

**Comment by asungomez** [comment]
There are some small workarounds for this. You can't prevent `amplify push` from overwriting the template, but you can trick it to generate the stack you want, to some extent.

I've found that adding to both `backend-config.json` and `amplify-meta.json` files some data manually (you need to duplicate it in both), and then running `amplify push --force`, makes several changes to the container stack. Specifically:

## Add output parameters from another stack as input parameters

You need to update the `dependsOn` array and add the input parameters you want in the following format:

```json
{
  "category": "",
  "resourceName": <Name of your stack>,
  "attributes": [
    <Name of attribute 1>,
    <Name of attribute 12>,
    ...
  ]
}
```

For example, if you have a container-based API called `"MyContainerApi"` and a custom resource called `"MyCustomResource"` that outputs the attributes `"CustomAttribute1"` and `"CustomAttribute2"`, you `backend-config.json` and `amplify-meta.json` files will initially look similar to this:

```json
{
  "api": {
    "MyContainerApi": {
      "apiType": "REST",
      "build": true,
      "dependsOn": [
        {
          "attributes": [
            "ClusterName",
            "VpcId",
            "VpcCidrBlock",
            "SubnetIds",
            "VpcLinkId",
            "CloudMapNamespaceId"
          ],
          "category": "",
          "resourceName": "NetworkStack"
        }
      ],
      "deploymentMechanism": "FULLY_MANAGED",
      "iamAccessUnavailable": true,
      "imageSource": {
        "type": "CUSTOM"
      },
      "providerPlugin": "awscloudformation",
      "resourceName": "graphql",
      "restrictAccess": false,
      "service": "ElasticContainer",
      "skipHashing": false
    }
  },
  "custom": {
    "MyCustomResource": {
      "dependsOn": [],
      "providerPlugin": "awscloudformation",
      "service": "customCloudformation"
    }
  }
}
```

And the stack name for your custom resource will be something like `"customMyCustomResource"` (you can check the name on your root stack in CF). Then you need to add the following object to the `dependsOn` array of your API:

```json
{
  "category": "",
  "resourceName": "customMyCustomResource", // Name of the stack
  "attributes": [
    // Name of the output attributes you want to input
    "CustomAttribute1",
    "CustomAttribute2"
  ],
}
```

So the file ends up looking like this:

```json
{
  "api": {
    "MyContainerApi": {
      "apiType": "REST",
      "build": true,
      "dependsOn": [
        {
          "attributes": [
            "ClusterName",
            "VpcId",
            "VpcCidrBlock",
            "SubnetIds",
            "VpcLinkId",
            "CloudMapNamespaceId"
          ],
          "category": "",
          "resourceName": "NetworkStack"
        },
        {
          "category": "",
          "resourceName": "customMyCustomResource",
          "attributes": ["CustomAttribute1", "CustomAttribute2"]
        }
      ],
      "deploymentMechanism": "FULLY_MANAGED",
      "iamAccessUnavailable": true,
      "imageSource": {
        "type": "CUSTOM"
      },
      "providerPlugin": "awscloudformation",
      "resourceName": "graphql",
      "restrictAccess": false,
      "service": "ElasticContainer",
      "skipHashing": false
    }
  },
  "custom": {
    "MyCustomResource": {
      "dependsOn": [],
      "providerPlugin": "awscloudformation",
      "service": "customCloudformation"
    }
  }
}
```

Then run `amplify push --force`, and you'll see the parameters in your API stack:

```json
"customMyCustomResourceCustomAttribute1": {
  "Type": "String"
},
"customMyCustomResourceCustomAttribute2": {
  "Type": "String"
},
```

## Using the parameters as environment variables for your container

Having the parameters is nice, but since you can't make any manual changes to the file, you can't really reference them, so they're kinda useless. I've discovered how to use them as environment variables for a container. You need to add the attribute `"environmentMap"` to your containerized API definition like this:

```json
{
  "environmentMap": {
    "CUSTOM_ATTRIBUTE_1": {
      "Ref": "customMyCustomResourceCustomAttribute1"
    },
    "CUSTOM_ATTRIBUTE_2": {
      "Ref": "customMyCustomResourceCustomAttribute2"
    }
  }
}
```

So the final definition will be:

```json
{
  "api": {
    "MyContainerApi": {
      "apiType": "REST",
      "build": true,
      "dependsOn": [
        {
          "attributes": [
            "ClusterName",
            "VpcId",
            "VpcCidrBlock",
            "SubnetIds",
            "VpcLinkId",
            "CloudMapNamespaceId"
          ],
          "category": "",
          "resourceName": "NetworkStack"
        },
        {
          "category": "",
          "resourceName": "customMyCustomResource",
          "attributes": ["CustomAttribute1", "CustomAttribute2"]
        }
      ],
      "environmentMap": {
        "CUSTOM_ATTRIBUTE_1": {
          "Ref": "customMyCustomResourceCustomAttribute1"
        },
        "CUSTOM_ATTRIBUTE_2": {
          "Ref": "customMyCustomResourceCustomAttribute2"
        }
      }
      "deploymentMechanism": "FULLY_MANAGED",
      "iamAccessUnavailable": true,
      "imageSource": {
        "type": "CUSTOM"
      },
      "providerPlugin": "awscloudformation",
      "resourceName": "graphql",
      "restrictAccess": false,
      "service": "ElasticContainer",
      "skipHashing": false
    }
  },
  "custom": {
    "MyCustomResource": {
      "dependsOn": [],
      "providerPlugin": "awscloudformation",
      "service": "customCloudformation"
    }
  }
}
```

You will see how the environment variables are created on your CF template, on the `containersDefinition` section of your Task:

```json
{
  "ContainerDefinitions": [
    {
      "Command": [],
      "EntryPoint": [],
      "Environment": [
        {
          "Name": "CUSTOM_ATTRIBUTE_1",
          "Value": {
            "Ref": "customMyCustomResourceCustomAttribute1"
          }
        },
        {
          "Name": "CUSTOM_ATTRIBUTE_2",
          "Value": {
            "Ref": "customMyCustomResourceCustomAttribute2"
          }
        }
      ],
      // ...rest of the container definition
    }
  ]
}
```
